Callbacks for some operations

mauricioszabo opened the pull request:
I'm working on a plugin that'll integrate with Proto-Repl: https://github.com/mauricioszabo/clojure-plus

I don't think most of the functionality I'm adding on clojure-plus is necessary here on proto-repl, so I'm implementing it on another place. But, to be able to implement it, I need to find out when proto-repl is ready to receive commands, and when it's not. So, I've added three simple callbacks in Proto-Repl.

Please, feel free to ask me any questions you might have :-)

mauricioszabo committed:
WIP

mauricioszabo committed:
Merge remote-tracking branch 'original/master' into callbacks

jasongilman commented:
That seems like a reasonable change to make. I'm interested in what kind of features you're putting into Clojure plus. 

Could I ask you make a few small changes as part of the pull request?
- Add comments to the new functions you added that these are for third party integration.
- Add some comments to https://github.com/jasongilman/proto-repl/blob/master/extending_proto_repl.md about listening for Proto REPL events. It could be a new top level section.

mauricioszabo committed:
Add docs to subscribe to protoRepl events.

mauricioszabo commented:
I've added the docs. Don't know if they are good enough, through.

As for Clojure Plus, I'm mostly experimenting. First, I've changed the way of doing refresh (in my case, refresh was clearing `user` namespace and the next refresh was not running the correct code), and added two configs: one that would run _before_ every refresh, and one _after_ every refresh. I'm mostly doing this because I'm working with a project where they fire up a bunch of servers, then before each refresh I need some code to kill everything and after refresh I fire then again. I've added too a config when I would decide when to refresh (after each save, after connecting, etc). There, I've got some problems with Midje (when refresh failed, sometimes midje lib messes everything, so in case of refresh failure I try to do a full refresh).

Second thing I've done is add a "serial exec" - for some reason (I don't really know why!) when we send commands to nREPL, it tries to run some codes in parallel. So, I've added a "promised repl", one that would serialize commands. Then, I've added code to simulate LightTable's "watch expressions" (without all the hackery in LT).

Now, I'm trying to meddle with namespace - resolve symbols, add imports, remove unused imports, these things. I've made a video (it's still unlisted) with the experiments I'm working right now: https://www.youtube.com/watch?v=-IH0xmCfysM&w=420&h=315

Let me know what do you think about it!

jasongilman commented:
That sounds cool. I haven't watched your video yet but I'd like to when I get a chance. Thanks for making the updates.
